### DISCUSSION TITLE
Auto updating GRUB via systemd at shutdown and reboot

### DISCUSSION THREAD

**Discussion by danboid** [Antynea/grub-btrfs#270] [General]
8th April 2023 EDIT

grub-btrfsd got refactored and fixed (for Debian users) on the 5th April so this guide is no longer necessary. You are recommended to configure and use grub-btrfsd as described in the grub-btrfs README instead.

The current instructions in the grub-btrfs README to configure the grub-btrfsd service to update GRUB when you create or delete snapshots don't work for me under Debian 11 or 12 so instead I have configured systemd to update GRUB every time I reboot or power off like so:

Create `/etc/systemd/system/update-grub.service`

```
[Unit]
Description=Update grub boot menu at shutdown
DefaultDependencies=no
Before=shutdown.target

[Service]
Type=oneshot
ExecStart=/etc/rc6.d/K99update-grub.sh

[Install]
WantedBy=shutdown.target
```

Create a new shutdown script `/etc/rc6.d/K99update-grub.sh` to contain:

```
#!/bin/bash
/usr/sbin/grub-mkconfig -o /boot/grub/grub.cfg
```

Make sure that the path for both for grub-mkconfig and its config file are correct for your distro. 
These paths should work for Debian and Ubuntu at least. Make the script executable:

```
$ sudo chmod ugo+x /etc/rc6.d/K99update-grub.sh
```

Then run:

```
$ sudo systemctl enable update-grub.service
```

I've been having issues with the latest git version of grub-btrfs not finding my timeshift btrfs snapshots so instead of running that I'm using the the latest Kali linux .deb package from http://http.kali.org/pool/main/g/grub-btrfs/ , currently http://http.kali.org/pool/main/g/grub-btrfs/grub-btrfs_4.12-0kali1~jan+nur1_all.deb

**Comment by Schievel1** [comment]
Thanks for that. We also had the same idea a while ago, see https://github.com/Antynea/grub-btrfs/issues/138#issuecomment-766918328. 
Openrc offers `/etc/local.d/`for that, see https://github.com/Antynea/grub-btrfs/blob/0182e119516a14d49778bec94d3aeb9ce5742a83/README.md#openrc

**Reply by danboid** [reply]
Would you like me to submit a PR for the grub-btrfs README to replace the current systemd suggestion with this, replacing the current systemd instructions for auto-updating the grub boot menu?

I first tried grub-btrfs a year or so ago and I managed to get the systemd dir monitoring feature to work to monitor my snapshots dir for updating grub but by using my own implementation (again, the README suggestion wasn't working back then either) but that code doesn't work any more (under Debian) so I presume something changed in systemd?

This method is simpler and hence is less likely to break. The main thing is that it actually works (for me at least) whilst the current README suggestion doesn't. I wrote this guide so that it should work under other systemd distros too, if the users check the paths are correct it should be fine.

**Comment by danboid** [comment]
@Schievel1 

I have updated this guide to note that grub-btrfsd now works for Debian users too (to auto update GRUB) so everyone should use that instead.
